src/js/background.js

```diff
@@ -0,0 +1,214 @@
+const setLoadingState = async (active, tabId) => {
+  const [calExist] = await browser.tabs.executeScript(tabId, {
+    code: `document.querySelector(".cal-loading") !== null;`,
+  });
+
+  if (!calExist) {
+    await browser.tabs.executeScript(tabId, {
+      file: "/src/js/createLoading.js",
+    });
+    await browser.tabs.insertCSS(tabId, { file: "/src/css/createLoading.css" });
+  }
+
+  await browser.tabs.executeScript(tabId, {
+    code: `document.querySelector(".cal-loading").style.display = ${active} ? "flex" : "none";`,
+  });
+};
+
+const sendLogMessage = async (message, tabId) => {
+  await browser.tabs.executeScript(tabId, {
+    code: `console.log("[Chess.com analyse at lichess]: ${message}");`,
+  });
+};
+
+const fetchJSON = async (url, tabId) => {
+  try {
+    const response = await fetch(url);
+    if (!response.ok)
+      throw new Error("Error when attempting to fetch resource.");
+    return response.json();
+  } catch (error) {
+    sendLogMessage(error, tabId);
+    return null;
+  }
+};
+
+/*
+ * Portable Game Notation (PGN) is a standard plain text format for
+ * recording chess games, which can be read by humans and is also
+ * supported by most chess software
+ */
+const getPGN = async (playerName, gameUrl, tabId) => {
+  try {
+    const archives = (
+      await fetchJSON(
+        `https://api.chess.com/pub/player/${playerName}/games/archives`,
+        tabId
+      )
+    ).archives;
+    /*
+     * Archives are given from the oldest to the most recet games.
+     * Thus iterating from the last one (the most recent) is usually
+     * better, because players usually want to analyse their most recent
+     * games. You will probably make less fetches
+     */
+    for (let i = archives.length - 1; i >= 0; i--) {
+      const games = (await fetchJSON(archives[i], tabId)).games;
+      for (let j = games.length - 1; j >= 0; j--) {
+        if (games[j].url === gameUrl) {
+          return games[j].pgn;
+        }
+      }
+    }
+  } catch (error) {
+    sendLogMessage(error, tabId);
+    return null;
+  }
+
+  return null;
+};
+
+const getPGNManual = async (tabId) => {
+  await waitAndClick(".share");
+  await waitAndClick(".board-tab-item-underlined-component");
+  await waitAndClick(".share-menu-tab-pgn-toggle");
+
+  if (await waitForElement("[name='pgn']")) {
+    const [pgn] = await browser.tabs.executeScript(tabId, {
+      code: `document.querySelector("[name='pgn']").value;`,
+    });
+    await waitAndClick("[data-cy='share-menu-close']");
+
+    return pgn;
+  }
+
+  return null;
+};
```

Comment:
I feel like this function is a good candidate to be a workflow to be driven by the content script, rather than the background page.

Right now we have the background page as the director of actions. It gives low-level instructions to individual tabs, but overall, the background script is in control. There is overhead in communicating between the realms (probably negligible, but still).

An alternative approach would be to have the background script send a high-level message to the content script, like `getPGNManual`. The content script would receive that message and perform the entire workflow, responding with some information back to the background script.

This feels more natural to me if there are multiple steps to be taken that need access to the DOM, like they do here.

This approach has downsides, though. The background script now needs to have some measures of detecting that the tab was closed and will never respond to a message. This is probably still a problem with the existing code, but is much smaller, as I assume some `waitAndClick` operation will throw an error if the tab was closed, so the background script won't get stuck.

I'm sharing an idea, not expecting anything to be changed.